### PR TITLE
Add final diagnostics, refactoring

### PR DESCRIPTION
**Pull request by j0shuams**
* add namespace diagnostics
* move all diagnostic code out of generator.cs
* use helper class for collecting all authored types
* add invalid type diagnostics
* check type symbol kinds instead of strings where possible
* split diagnostic code into two files for readability
* comment fixups
* no more pass by ref, no more returning bool to `found` var
* less generic functions
* more tests
* turn some one-off functions into local functions
* better signature checking algorithm 

* there is still one test (and a variation of it) failing still, but this might not be an issue
* authoring sample built once fine then complained about ReadOnlyArray / WriteOnlyArray attribute not being found, will try again 

I have a couple issues on the roslyn-sdk repo for things that seem like they should work
https://github.com/dotnet/roslyn-sdk/issues/678
https://github.com/dotnet/roslyn-sdk/issues/677

### CODE REVIEW

File: src/Authoring/DiagnosticTests/NegativeData.cs
Comment (j0shuams) [inline src/Authoring/DiagnosticTests/NegativeData.cs:6]: update comment

File: src/Authoring/DiagnosticTests/PositiveData.cs
Comment (j0shuams) [inline src/Authoring/DiagnosticTests/PositiveData.cs:5]: fixed

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:145]: "instead of in/out"

Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:145]: "instead of in/out"

File: src/Authoring/AuthoringSample/Program.cs
Comment (manodasanW) [inline src/Authoring/AuthoringSample/Program.cs:463]: We might need to improve our diagnostics to detect whether the function belongs to a custom type mapping.  In this case, this function is part of a custom mapping interface which means this function as is won't be projected.  But it would be removed and a mapping of it would be projected.

File: src/Authoring/AuthoringSample/Program.cs
Comment (manodasanW) [inline src/Authoring/AuthoringSample/Program.cs:473]: Same here, this function is part of a custom mapped interface.

File: src/Authoring/AuthoringSample/Program.cs
Comment (manodasanW) [inline src/Authoring/AuthoringSample/Program.cs:631]: see other comment.

File: src/Authoring/DiagnosticTests/PositiveData.cs
Comment (manodasanW) [inline src/Authoring/DiagnosticTests/PositiveData.cs:39]: Isn't this a positive test case file?

Comment (j0shuams) [reply src/Authoring/DiagnosticTests/PositiveData.cs:39]: Yes this is a valid test, the comment is a copy-paste error, fixed in last commit

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:107]: Technically the base type can have another base type.  Does that matter in the scenarios this is used in?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:109]: A type can have a base type and also implement other interfaces.  Should we check interfaces if this is false?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:199]: What if it is A.B.C.type and A.B.WinMD or something like that.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:352]: Should we normalize the types as technically they can already include System and then do Linq...

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:343]: Don't need to address right now, but we need to do a not valid type check rather than an invalid type check.  But can be addressed later on scenarios that we identify if not already handled.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:188]: What about 2 namespaces defined multiple times with the same case?  Is that a scenario to consider here?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:309]: To confirm, is this looking for a partial match due to contains?

File: src/Authoring/WinRT.SourceGenerator/Generator.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:182]: Consider WinRTComponentScanner or WinRTComponentDiagnostics?

Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs:182]: I like it! Thanks

File: src/Authoring/WinRT.SourceGenerator/Generator.cs
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:189]: take this out before final merge, can we use a directive here like `#if NET5_0`

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:107]: It seems like there is some ambiguity about the `BaseType` field of type symbols. For type symbols of classes, any interface implemented is actually a basetype (e.g. IAsyncAction is not in the symbol's `AllInterfaces` field, but in its `BaseType` field).

For an interface type symbol, this is not true, and the `IAsyncAction` symbol **is** in the `AllInterfaces` field. So the loop is for interfaces, when we want to check what a class implements, we need it's `BaseType`, but an interface has a null `BaseType`. 

And class type symbols need a `BaseType` check, to find `System.Exception` for example. 

It is also curious that calling `GetTypeByMetadataName("Windows.Foundation.IAsyncAction")` returns `null` instead of the type symbol we might expect. This is why we have to do the `==` check.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:109]: I had moved the interface loop up to before we check the base type. Not sure what you mean by check interfaces if this is false

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:352]: I am, I build up a string of the containing namespace(s) for the type before comparing it to these strings. I don't have a test for if a user defined a class named Enumerable and uses it, so I will add that

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:343]: Went ahead and changed the name, but not sure I understand the difference in not valid and invalid

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:188]: If the namespace name is valid (e.g. is the assembly name or a namespace prefix with the assembly name) then that's fine. Otherwise it will get the "disjoint namespace" diagnostic, meaning the types won't be accessible. I'll add a test for this thanks

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:309]: I found an alternative approach that doesn't use this kind of check

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:199]: Yes I have tests for these and have added one that more obviously tests what you are saying -- they are blocked by issue https://github.com/dotnet/roslyn/issues/50112 

As the namespace C will have "{global namespace}" as its ContainingNamespace, instead of "A.B"

File: src/Authoring/WinRT.SourceGenerator/Generator.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs:189]: tried this and setting it in the diagnostictest project properties didn't seem to apply, so it didn't work...

File: src/Authoring/AuthoringSample/Program.cs
Comment (j0shuams) [reply src/Authoring/AuthoringSample/Program.cs]: I've updated it to now look for IList, IDictionary, and IReadOnlyDictionary, and filter out these interfaces' methods and properties from those that get scanned

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:373]: You only need the interfaces as those are the ones that can be implemented by component classes.  The custom mapped classes I believe are all sealed so they are not used for extending.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:377]: Any way to make the method lookup generic.  I am concerned about overloaded methods with different # of parameters that are from the component.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:422]: Same here.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticRules.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticRules.cs]: Did you plan on addressing this or future PR?

Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticRules.cs]: Yes we want to do it now, thanks

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs]: Addressed

File: src/Authoring/WinRT.SourceGenerator/DiagnosticRules.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticRules.cs]: addressed

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:413]: addressed

File: src/Authoring/WinRT.SourceGenerator/Generator.cs
Comment (j0shuams) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:188]: when we merge this pr, i will take these out before merging the main topic pr

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:50]: Is this intentional?  Do you plan on going back to Equals as is?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:370]: Any chance for both this list and the other, we can use the one already in WinRTTypeWriter.cs.  We can move it to a helper file if needed.  Is there any difference between both lists other than this only has the keys?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:422]: Should the other 4 types here be qualified names similar to the first one?

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:126]: I feel like you would want a symbol comparison here as what happens if 2 functions have same name, but have a diff # of parameters and one is from a custom type mapping while the other isn't.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:50]: I went back to the original, thanks for catching this.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:370]: yes we are sharing now

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:422]: Well gettypebymetadataname fails for the qualified ones, and I checked one of our tests for these interfaces, and when the class implements it, the interface's containing namespace is global (unless qualified when giving the class the base type), which means qualified names won't help, just metadata name.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs]: addressed

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:119]: might be performant to do a loop over interface members and build a list of class members using FindImplementationForInterfaceMember and then do a remove all or where not contains.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:371]: My only concern was if someone used the full typename or somewhere whether we will miss it.  But if we are comparing metadata name, it probably is fine. Just seemed inconsistent with System.Exception.  And what happens if someone had their own version of IAsyncAction in their own namespace?

**Review by manodasanW — APPROVED** [review]

File: src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:371]: Yeah we do catch full name, but I bet we won't let it happen if they write IAsyncAction, but should they be allowed to write IAsyncAction? they are are encouraged (in windmdexp) to use IAsyncInfo class for exception code

Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:371]: Well I meant IAsyncAction from some other namespace (different type).

Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:371]: I meant is Exception a valid winrt type? implementing IAsyncInfo would mean having a property that is an Exception.. Maybe there's an edge case there with them also writing their own Exception class?

Comment (manodasanW) [reply src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs:371]: I think it is valid, but probably want to confirm.  You probably also want to check if inheriting is allowed, but I think not.

**Pull request merged**
